src/java_bytecode/java_types.cpp

```diff
@@ -167,6 +167,218 @@ exprt java_bytecode_promotion(const exprt &expr)
+                                                  size_t starting_point = 0);
+
+/// Given a substring of a descriptor or signature that contains 1 or more types
+/// parse out the inidividual types. This is used for parsing the parameters of
+/// a function or the gneeric arguments contained within angle brackets.
+/// \param src: The input string that is wrapped in either ( ) or < >
+/// \param class_name_prefix: The name of the class to use to prefix any found
+///   generic types
+/// \param opening_bracket: For checking string is passed in as expected, the
+///   opening bracket (i.e. '(' or '<').
+/// \param closing_bracket: For checking string is passed in as expected, the
+///   closing bracket (i.e. ')' or '>').
+/// \return A vector of types that the string represents
+std::vector<typet> parse_list_types(const std::string src,
+                                    const std::string class_name_prefix,
+                                    const char opening_bracket,
+                                    const char closing_bracket) {
+  PRECONDITION(src[0] == opening_bracket);
+  PRECONDITION(src[src.size() - 1] == closing_bracket);
+  std::vector<typet> type_list;
+  for (std::size_t i = 1; i < src.size() - 1; i++) {
+    size_t start = i;
+    while (i < src.size()) {
+      // parameter is an object type or instantiated generic type
+      if (src[i] == 'L') {
+        i = find_closing_semi_colon_for_reference_type(src, i);
+        break;
+      }
+
+      // parameter is an array
+      else if (src[i] == '[')
+        i++;
+
+      // parameter is a type variable
+      else if (src[i] == 'T')
+        i = src.find(';', i); // ends on ;
+
+      // type is an atomic type (just one character)
+      else {
+        break;
+      }
+    }
+
+    std::string sub_str = src.substr(start, i - start + 1);
+    const typet &new_type = java_type_from_string(sub_str, class_name_prefix);
+    INVARIANT(new_type != nil_typet(), "Failed to parse type");
+
+    if (new_type.id() == ID_symbol) {
+      UNREACHABLE;
+    }
+
+    type_list.push_back(new_type);
+  }
+  return type_list;
+}
+
+/// For parsing a class type reference
+/// \param src: The input string
+///   Either a signature: "Lpackage/class<TT;>.innerclass<Ljava/lang/Foo;>;
+///   Or a descriptor: "Lpackage.class$inner;
+/// \param class_name_prefix: The name of the class to use to prefix any found
+///   generic types
+/// \return The reference type if parsed correctly, no value if parsing fails.
+optionalt<reference_typet>
+build_class_name(const std::string src, const std::string &class_name_prefix) {
+  PRECONDITION(src[0] == 'L');
+  // ends on ;
+  if (src[src.size() - 1] != ';')
+    return optionalt<reference_typet>();
+
+  std::string container_class = gather_full_class_name(src);
+  std::string identifier = "java::" + container_class;
+  symbol_typet symbol_type(identifier);
+  symbol_type.set(ID_C_base_name, container_class);
+
+  std::size_t f_pos = src.find('<', 1);
```

Comment:
Same loop rotation recommendation as above -- prefer writing this update condition once and doing `while(true) { .... if(exit_condition) break; ... }` over duplicating it like this

Reply:
See above discussion, will apply both changes if you want, otherwise neither